Mathlib/Analysis/Normed/Affine/AsymptoticCone.lean

```diff
@@ -0,0 +1,428 @@
+    eta_expand
+    simp_rw [asymptoticNhds_eq_smul_vadd _ p, vadd_pure]
+    nth_rw 2 [← nhds_bind_nhds]
+    intro s ⟨t₁, ht₁, t₂, ht₂, hs⟩
+    rw [mem_bind] at ht₂ ⊢
+    obtain ⟨t₃, ht₃, ht₂⟩ := ht₂
+    exact ⟨t₃, ht₃, fun u hu => ⟨t₁, ht₁, t₂, ht₂ u hu, hs⟩⟩
+  · rw [← pure_bind v (asymptoticNhds k P)]
+    exact bind_mono (pure_le_nhds v) .rfl
+
+theorem asymptoticNhds_bind_nhds {v : V} :
+    (asymptoticNhds k P v).bind 𝓝 = asymptoticNhds k P v := by
+  refine le_antisymm (fun s h => ?_) (bind_mono le_rfl (.of_forall pure_le_nhds))
+  have ⟨p⟩ : Nonempty P := inferInstance
+  rw [asymptoticNhds_eq_smul_vadd _ p, vadd_pure] at h ⊢
+  rw [← nhds_bind_nhds] at h
+  obtain ⟨t₁, ht₁, t₂, ht₂, hs⟩ := h
+  rw [mem_bind] at ht₂
+  obtain ⟨t₃, ht₃, ht₂⟩ := ht₂
+  rw [bind_map, mem_bind]
+  refine ⟨(t₁ ∩ Set.Ioi 0) • t₃, smul_mem_smul (inter_mem ht₁ (Ioi_mem_atTop _)) ht₃,
+    Set.forall_mem_image2.mpr fun c ⟨hc₁, hc₂⟩ u hu => ?_⟩
+  rw [show s = (· -ᵥ p) ⁻¹' ((· +ᵥ p) ⁻¹' s) by simp [Set.preimage_preimage]]
+  apply tendsto_id.vsub tendsto_const_nhds
+  rw [vadd_vsub]
+  filter_upwards [smul_mem_nhds_smul₀ hc₂.ne' (ht₂ u hu)]
+  rw [← Set.image_smul, Set.forall_mem_image]
+  exact fun w hw => hs (Set.smul_mem_smul hc₁ hw)
+
+theorem asymptoticNhds_bind_asymptoticNhds {v : V} :
+    (asymptoticNhds k V v).bind (asymptoticNhds k P) = asymptoticNhds k P v := by
+  refine Filter.ext' fun p => ?_
+  rw [asymptoticNhds_eq_smul, eventually_bind, ← map₂_smul, ← map_prod_eq_map₂, eventually_map,
+    ← nhds_bind_asymptoticNhds, eventually_bind]
+  nth_rw 2 [← map_snd_prod (atTop (α := k)) (𝓝 v)]
+  rw [eventually_map]
+  apply eventually_congr
+  filter_upwards [tendsto_fst.eventually (eventually_gt_atTop 0)] with ⟨c, u⟩ (hc : 0 < c)
+  simp only [asymptoticNhds_smul _ hc]
+
+variable (k) in
+
+/-- The set of directions `v` for which the set has points arbitrarily far in directions near `v`.
+-/
+def asymptoticCone (s : Set P) : Set V := {v | ∃ᶠ p in asymptoticNhds k P v, p ∈ s}
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem mem_asymptoticCone_iff {v : V} {s : Set P} :
+    v ∈ asymptoticCone k s ↔ ∃ᶠ p in asymptoticNhds k P v, p ∈ s :=
+  Iff.rfl
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+@[simp]
+theorem asymptoticCone_empty : asymptoticCone k (∅ : Set P) = ∅ :=
+  Set.eq_empty_iff_forall_notMem.mpr fun _ => frequently_false _
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+@[mono]
+theorem asymptoticCone_mono {s t : Set P} (h : s ⊆ t) : asymptoticCone k s ⊆ asymptoticCone k t :=
+  fun _ h' => h'.mono h
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem asymptoticCone_union {s t : Set P} :
+    asymptoticCone k (s ∪ t) = asymptoticCone k s ∪ asymptoticCone k t := by
+  ext
+  simp only [Set.mem_union, mem_asymptoticCone_iff, Filter.frequently_or_distrib]
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem Set.Finite.asymptoticCone_biUnion {ι : Type*} {s : Set ι} (hs : s.Finite) (f : ι → Set P) :
```

Comment:
The dot notation doesn't help disambiguate anything here, so I would drop it
```suggestion
theorem asymptoticCone_biUnion {ι : Type*} {s : Set ι} (hs : s.Finite) (f : ι → Set P) :
```
same below